CASSANDRA-17692 June 2022 blog "Apache Cassandra 4.1: New SSTable Identifiers" (#141)
apache/cassandra-website: site-content/source/modules/ROOT/pages/blog/Apache-Cassandra-4.1-New-SSTable-Identifiers.adoc

= Apache Cassandra 4.1: New SSTable Identifiers
:page-layout: single-post
:page-role: blog-post
:page-post-date: June 16, 2022
:page-post-author: Jacek Lewandowski

> :description: SSTable Identifiers in Apache Cassandra 4.1

Reviewer: ```suggestion
:description: New SSTable Identifiers in Apache Cassandra 4.1
```

:keywords: apache cassandra, 4.1, sstable

:!figure-caption:

.Image credit: https://unsplash.com/@qwitka[Maksym Kaharlytskyi on Unsplash^]
image::blog/apache-cassandra-4.1-new-sstable-identifiers-unsplash-maksym-kaharlytskyi.jpg[SSTable Identifiers in Apache Cassandra 4.1]

Apache Cassandra, like many other databases, stores data in files. These files are located in data directories and organized in SSTables. This post will discuss the directory layout and the naming pattern used for these files and explain the new naming pattern introduced in Apache Cassandra 4.1.

=== SSTables

> SSTables are files where Cassandra stores data from tables. In a typical operation, an SSTable is created either as a result of flushing a http://distributeddatastore.blogspot.com/2020/03/cassandra-memtable.html[memtable to disk^] or a https://cassandra.apache.org/doc/latest/cassandra/operating/compaction/index.html[compaction process]. Each SSTable contains data from a single table, but for a single table, there are usually many SSTables.

Reviewer: Fix `/doc` link:
```suggestion
SSTables are files where Cassandra stores data from tables. In a typical operation, an SSTable is created either as a result of flushing a http://distributeddatastore.blogspot.com/2020/03/cassandra-memtable.html[memtable to disk^] or a link:/doc/latest/cassandra/operating/compaction/index.html[compaction process]. Each SSTable contains data from a single table, but for a single table, there are usually many SSTables.
```

A single SSTable is made of multiple files, called components. These components are generally specific to the SSTable format. BigTable is the only format supported right now, and they are the only type of component you will see being created by Apache Cassandra (at least at the time of writing). For example, a single SSTable can be a set of such files:

> ----

Reviewer: ``````suggestion
```
``````

    nb-1-big-CompressionInfo.db
    nb-1-big-Data.db
    nb-1-big-Digest.crc32
    nb-1-big-Filter.db
    nb-1-big-Index.db
    nb-1-big-Statistics.db
    nb-1-big-Summary.db
    nb-1-big-TOC.txt

> ----

Reviewer: ``````suggestion
```
``````

> You can read more about the particular SSTable components of the BigTable format in the  https://cassandra.apache.org/doc/latest/cassandra/architecture/storage_engine.html#sstables[documentation].

Reviewer: Fix `/doc` link:
```suggestion
You can read more about the particular SSTable components of the BigTable format in the link:/doc/latest/cassandra/architecture/storage_engine.html#sstables[documentation^].
```

=== Directory Layout and File Names

SSTable files are stored in data directories. The directory layout consists of a directory per keyspace and a directory per table under the keyspace directory. 

> ----

Reviewer: ``````suggestion
```
``````

data0/
  /ks_foo
    /tab_bar-<id>
   	/<version>-<generation id>-<format>-<component>.<ext>

> 

Reviewer: ```suggestion
```

data1/
  /ks_foo
    /tab_bar-<id>
   	/<version>-<generation id>-<format>-<component>.<ext>

> ----

Reviewer: ``````suggestion
```
``````

*NOTE:* Cassandra includes live data directories and backup directories but ignores snapshots directories when performing its startup scan. Therefore, there may be SSTables with the same identifier among all the data directories while being different SSTables. 
The general identifiers based on the natural numbers aim to be unique per Cassandra node and table. This means that not only two SSTables of two different tables created on the same node may have the same identifiers and, thus, the same file names, but two different SSTables of the same table created on different nodes.

As you might expect, there can be some maintenance problems due to the identifier properties discussed above. For example, as we illustrate below, truncation of a table triggers snapshot creation and removal of all SSTables from data directories. If the node is restarted and there is no SSTable created before that, the sequence is restarted from the beginning because there is no existing SSTable for identifying the last generated identifier:

> ----

Reviewer: ``````suggestion
```
``````

  /ks_foo
    /tab_bar-<id>
      /nb-1-big-Data.db

> ----

Reviewer: ``````suggestion
```
``````

There is a snapshot made before the truncation - that is, Cassandra creates hard links to all the SSTables files in a snapshot directory then files are removed from the live data directory:

> ----

Reviewer: ``````suggestion
```
``````

  /ks_foo
    /tab_bar-<id>
      /snapshots
        /truncated-<timestamp>-tab_bar
          /nb-1-big-Data.db

> ----

Reviewer: ``````suggestion
```
``````

When the node gets restarted, Cassandra forgets about the current sequence and starts it over; when a new SSTable is stored, it gets ‘1’ as the identifier:

> ----

Reviewer: ``````suggestion
```
``````

  /ks_foo
    /tab_bar-<id>
      /nb-1-big-Data.db
      /snapshots
        /truncated-<timestamp>-tab_bar
          /nb-1-big-Data.db

> ----

Reviewer: ``````suggestion
```
``````

As you can see, it is possible to have two SSTables with the same name but with potentially different content. This situation only becomes a problem when a user stores the SSTables in a different location for a backup. The backups of the SSTables will likely clash with the existing ones due to the identical file names.

=== Introducing Globally Unique Identifiers

To solve some of the problems with SSTable identifiers based on natural numbers, Cassandra 4.1 introduces the ability to switch to globally unique identifiers. These new identifiers are based on Time UUIDs (UUID type 1), though their string representation is different, making them lexically ordered and providing a little aid for the administrators. 
The structure of a globally unique identifier is as follows:

> `<date part>_<time part>_<nano part><random part>`

Reviewer: ``````suggestion
```
<date part>_<time part>_<nano part><random part>
```
``````
